onnxruntime/core/providers/coreml/coreml_execution_provider.cc

```diff
@@ -0,0 +1,314 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "coreml_execution_provider.h"
+
+#include "core/framework/allocatormgr.h"
+#include "core/framework/compute_capability.h"
+#include "core/graph/graph_viewer.h"
+#include "core/session/onnxruntime_cxx_api.h"
+
+#include "model/model.h"
+#include "model/host_utils.h"
+#include "builders/helper.h"
+#include "builders/model_builder.h"
+
+namespace onnxruntime {
+
+constexpr const char* COREML = "CoreML";
+
+CoreMLExecutionProvider::CoreMLExecutionProvider()
+    : IExecutionProvider{onnxruntime::kCoreMLExecutionProvider} {
+  AllocatorCreationInfo device_info(
+      [](int) {
+        return onnxruntime::make_unique<CPUAllocator>(OrtMemoryInfo(COREML, OrtAllocatorType::OrtDeviceAllocator));
+      });
+
+  InsertAllocator(CreateAllocator(device_info));
+
+  AllocatorCreationInfo cpu_memory_info(
+      [](int) {
+        return onnxruntime::make_unique<CPUAllocator>(
+            OrtMemoryInfo(COREML, OrtAllocatorType::OrtDeviceAllocator, OrtDevice(), 0, OrtMemTypeCPUOutput));
+      });
+
+  InsertAllocator(CreateAllocator(cpu_memory_info));
+}
+
+CoreMLExecutionProvider::~CoreMLExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+CoreMLExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_view,
+                                       const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+
+  // We do not run CoreML EP on subgraph, instead we cover this in the control flow nodes
+  if (graph_view.IsSubgraph()) {
+    return result;
+  }
+
```

Comment:
Not following this. 

The EP doesn't need to recurse into subgraphs as the partitioning will do that, but just because it's currently in a subgraph I wouldn't have expect it to return nothing for GetCapability. It can still handle nodes in the subgraph can't it?

Reply:
Yes, the EP should be able to handle subgraph, but with this we may not have any perf gain or could get even worse perf, put this as pending investigation

---
In reply to: [563458222](https://github.com/microsoft/onnxruntime/pull/6392#discussion_r563458222) [](ancestors = 563458222)

Reply:
That's the same for any model though isn't it - that perf could be worse if you use the EP? If the subgraph is large, we could get a perf win. I guess we need a config setting to enable/disable processing subgraphs so the user can test out with/without that enabled.

Reply:
Adding config is a pending work item, will try to handle that in a later PR

---
In reply to: [565059375](https://github.com/microsoft/onnxruntime/pull/6392#discussion_r565059375) [](ancestors = 565059375)